bitcoin/src/crypto/sighash.rs

```diff
@@ -64,24 +66,659 @@ sha256t_hash_newtype! {
+        (**self).extra_at(index)
+    }
+}
+
+impl<T: AsExtendedTransaction + ?Sized> AsExtendedTransaction for alloc::sync::Arc<T> {
+    type Extra = T::Extra;
+
+    fn prev_output_at(&self, index: usize) -> Option<&TxOut> {
+        (**self).prev_output_at(index)
+    }
+
+    fn extra_at(&self, index: usize) -> Option<&Self::Extra> {
+        (**self).extra_at(index)
+    }
+}
+
+impl AsTransaction for Transaction {
+    fn inputs_len(&self) -> usize {
+        self.input.len()
+    }
+
+    fn input_at(&self, index: usize) -> Result<&TxIn, transaction::InputsIndexError> {
+        self.tx_in(index)
+    }
+
+    fn outputs_len(&self) -> usize {
+        self.output.len()
+    }
+
+    fn output_at(&self, index: usize) -> Result<&TxOut, transaction::OutputsIndexError> {
+        self.tx_out(index)
+    }
+
+    fn version(&self) -> transaction::Version {
+        self.version
+    }
+
+    fn lock_time(&self) -> absolute::LockTime {
+        self.lock_time
+    }
+}
+
+impl AsMutTransaction for Transaction {
+    fn mut_input_at(&mut self, index: usize) -> Result<&mut TxIn, transaction::InputsIndexError> {
+        self.tx_in_mut(index)
+    }
+}
+
+impl<T> AsTransaction for transaction::extended::ExtendedTransaction<T> {
+    fn inputs_len(&self) -> usize {
+        self.inputs.len()
+    }
+
+    fn input_at(&self, index: usize) -> Result<&TxIn, transaction::InputsIndexError> {
+        self.tx_in(index).map(|input| &input.txin)
+    }
+
+    fn outputs_len(&self) -> usize {
+        self.outputs.len()
+    }
+
+    fn output_at(&self, index: usize) -> Result<&TxOut, transaction::OutputsIndexError> {
+        self.tx_out(index)
+    }
+
+    fn version(&self) -> transaction::Version {
+        self.version
+    }
+
+    fn lock_time(&self) -> absolute::LockTime {
+        self.lock_time
+    }
+}
+
+impl<T> AsMutTransaction for transaction::extended::ExtendedTransaction<T> {
+    fn mut_input_at(&mut self, index: usize) -> Result<&mut TxIn, transaction::InputsIndexError> {
+        self.tx_in_mut(index).map(|input| &mut input.txin)
+    }
+}
+
+impl<T> AsExtendedTransaction for transaction::extended::ExtendedTransaction<T> {
+    type Extra = T;
+
+    fn prev_output_at(&self, index: usize) -> Option<&TxOut> {
+        self.inputs.get(index).map(|input| &input.previous_output)
+    }
+
+    fn extra_at(&self, index: usize) -> Option<&Self::Extra> {
+        self.inputs.get(index).map(|input| &input.extra)
+    }
+}
+
+impl<T> AsMutExtendedTransaction for transaction::extended::ExtendedTransaction<T> {
+    fn mut_extra_at(&mut self, index: usize) -> Option<&mut Self::Extra> {
+        self.inputs.get_mut(index).map(|input| &mut input.extra)
+    }
+}
+
+trait AsTransactionExt: AsTransaction {
+    fn inputs(&self) -> Inputs<'_, Self> {
+        Inputs(self)
+    }
+
+    fn outputs(&self) -> Outputs<'_, Self> {
+        Outputs(self)
+    }
+}
+
+impl<T: AsTransaction + ?Sized> AsTransactionExt for T {}
```

Comment:
We can't because not all things implement all the features. E.g. `Transaction` cannot give you previous output.

Reply:
Ok, perhaps I'm missing some understanding of Rust here.
```rust
trait AsTransactionExt: AsTransaction {
```
Says that `AsTransactionExt` requires the trait `AsTransaction` to be implemented.

but
```rust
impl<T: AsTransaction + ?Sized> AsTransactionExt for T {}
```
Implements `AsTransactionExt` for any type that implements `AsTransaction`.

I moved the two trait methods from `AsTransactionExt` to `AsTransaction` and the code builds.

Reply:
Crap, I mistook it for `AsExtendedTransaction`.

The trait exists for convenience and it's an internal API. Notably, even if we wanted to publish it, it'd still make more sense to make it extended so that people won't mess up indices. Unless we want to make them implement iterators with GAT and all other mess. IMO not worth it.